// @antv/g2/lib/mark/rect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Rect = void 0;
const utils_1 = require("./utils");
const Rect = () => {
    return (index, scale, value, coordinate) => {
        const { x: X, x1: X1, y: Y, y1: Y1 } = value;
        const P = Array.from(index, (i) => {
            const p1 = [+X[i], +Y[i]];
            const p2 = [+X1[i], +Y[i]];
            const p3 = [+X1[i], +Y1[i]];
            const p4 = [+X[i], +Y1[i]];
            return [p1, p2, p3, p4].map((d) => coordinate.map(d));
        });
        return [index, P];
    };
};
exports.Rect = Rect;
const shapes = ['rect', 'hollow'];
exports.Rect.props = {
    defaultShape: 'rect',
    defaultLabelShape: 'label',
    composite: false,
    channels: [
        ...(0, utils_1.baseGeometryChannels)({ shapes }),
        { name: 'x', required: true },
        { name: 'y', required: true },
    ],
    preInference: [...(0, utils_1.basePreInference)(), { type: 'maybeZeroY1' }],
    postInference: [...(0, utils_1.basePostInference)(), ...(0, utils_1.tooltip1d)()],
    interaction: {
        shareTooltip: true,
    },
};
