// @antv/g2/lib/mark/text.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Text = void 0;
const utils_1 = require("./utils");
const Text = (options) => {
    const { cartesian = false } = options;
    if (cartesian)
        return utils_1.visualMark;
    return ((index, scale, value, coordinate) => {
        if (cartesian)
            return (0, utils_1.visualMark)(index, scale, value, coordinate);
        const { x: X, y: Y } = value;
        const offset = (0, utils_1.createBandOffset)(scale, value, options);
        const P = Array.from(index, (i) => {
            const p = [+X[i], +Y[i]];
            return [coordinate.map(offset(p, i))];
        });
        return [index, P];
    });
};
exports.Text = Text;
const shapes = ['text', 'badge'];
exports.Text.props = {
    defaultShape: 'text',
    defaultLabelShape: 'label',
    composite: false,
    channels: [
        ...(0, utils_1.baseGeometryChannels)({ shapes }),
        { name: 'x', required: true },
        { name: 'y', required: true },
        { name: 'text', scale: 'identity' },
        { name: 'fontSize' },
        { name: 'rotate' },
    ],
    preInference: [
        ...(0, utils_1.basePreInference)(),
        { type: 'maybeTuple' },
        { type: 'maybeVisualPosition' },
    ],
    postInference: [...(0, utils_1.basePostInference)(), ...(0, utils_1.tooltip2d)()],
};
