// @antv/gui/lib/core/gui.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GUI = void 0;
var tslib_1 = require("tslib");
var g_1 = require("@antv/g");
var util_1 = require("../util");
var GUI = /** @class */ (function (_super) {
    tslib_1.__extends(GUI, _super);
    function GUI(config) {
        return _super.call(this, config) || this;
    }
    GUI.prototype.connectedCallback = function () {
        this.render(this.attributes, this);
        this.bindEvents(this.attributes, this);
    };
    GUI.prototype.update = function (cfg, animate) {
        var _a;
        if (cfg === void 0) { cfg = {}; }
        this.attr((0, util_1.deepAssign)({}, this.attributes, cfg));
        return (_a = this.render) === null || _a === void 0 ? void 0 : _a.call(this, this.attributes, this, animate);
    };
    GUI.prototype.clear = function () {
        this.removeChildren();
    };
    GUI.prototype.destroy = function () {
        this.removeAllEventListeners();
        this.removeChildren();
        this.remove();
    };
    GUI.prototype.attributeChangedCallback = function () { };
    GUI.prototype.bindEvents = function (attributes, container) { };
    return GUI;
}(g_1.CustomElement));
exports.GUI = GUI;
