// @antv/g2/lib/shape/interval/color.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Color = exports.rect = void 0;
const g_1 = require("@antv/g");
const d3_shape_1 = require("d3-shape");
const coordinate_1 = require("../../utils/coordinate");
const selection_1 = require("../../utils/selection");
const vector_1 = require("../.././utils/vector");
const utils_1 = require("../utils");
// Render rect in different coordinate.
function rect(points, value, coordinate, style = {}) {
    const { inset = 0, radius = 0, insetLeft = inset, insetTop = inset, insetRight = inset, insetBottom = inset, radiusBottomLeft = radius, radiusBottomRight = radius, radiusTopLeft = radius, radiusTopRight = radius } = style, rest = __rest(style, ["inset", "radius", "insetLeft", "insetTop", "insetRight", "insetBottom", "radiusBottomLeft", "radiusBottomRight", "radiusTopLeft", "radiusTopRight"]);
    if (!(0, coordinate_1.isPolar)(coordinate) && !(0, coordinate_1.isHelix)(coordinate)) {
        const tpShape = !!(0, coordinate_1.isTranspose)(coordinate);
        const [p0, , p2] = tpShape ? (0, utils_1.reorder)(points) : points;
        const [x, y] = p0;
        const [width, height] = (0, vector_1.sub)(p2, p0);
        // Deal with width or height is negative.
        const absX = width > 0 ? x : x + width;
        const absY = height > 0 ? y : y + height;
        const absWidth = Math.abs(width);
        const absHeight = Math.abs(height);
        const finalX = absX + insetLeft;
        const finalY = absY + insetTop;
        const finalWidth = absWidth - (insetLeft + insetRight);
        const finalHeight = absHeight - (insetTop + insetBottom);
        return (0, selection_1.select)(new g_1.Rect({}))
            .style('x', finalX)
            .style('y', finalY)
            .style('width', finalWidth)
            .style('height', finalHeight)
            .style('radius', [
            radiusTopLeft,
            radiusTopRight,
            radiusBottomRight,
            radiusBottomLeft,
        ])
            .call(utils_1.applyStyle, rest)
            .node();
    }
    // Render path in polar coordinate.
    const { y, y1 } = value;
    const center = coordinate.getCenter();
    const arcObject = (0, utils_1.getArcObject)(coordinate, points, [y, y1]);
    const path = (0, d3_shape_1.arc)()
        .cornerRadius(radius)
        .padAngle((inset * Math.PI) / 180);
    return (0, selection_1.select)(new g_1.Path({}))
        .style('path', path(arcObject))
        .style('transform', `translate(${center[0]}, ${center[1]})`)
        .style('radius', radius)
        .style('inset', inset)
        .call(utils_1.applyStyle, rest)
        .node();
}
exports.rect = rect;
/**
 * Render rect in different coordinate.
 * Calc arc path based on control points directly rather startAngle, endAngle, innerRadius,
 * outerRadius. This is not accurate and will cause bug when the range of y scale is [1, 0]
 * for cell geometry.
 */
const Color = (options) => {
    // Render border only when colorAttribute is stroke.
    const { colorAttribute, opacityAttribute = 'fill', first = true, last = true } = options, style = __rest(options, ["colorAttribute", "opacityAttribute", "first", "last"]);
    return (points, value, coordinate, theme) => {
        const { mark, shape, defaultShape } = value;
        const _a = (0, utils_1.getShapeTheme)(theme, mark, shape, defaultShape), { defaultColor, radius: defaultRadius = 0 } = _a, defaults = __rest(_a, ["defaultColor", "radius"]);
        const defaultLineWidth = defaults.lineWidth || 1;
        const { stroke, radius = defaultRadius, radiusTopLeft = radius, radiusTopRight = radius, radiusBottomRight = radius, radiusBottomLeft = radius, innerRadius = 0, innerRadiusTopLeft = innerRadius, innerRadiusTopRight = innerRadius, innerRadiusBottomRight = innerRadius, innerRadiusBottomLeft = innerRadius, lineWidth = colorAttribute === 'stroke' || stroke ? defaultLineWidth : 0, inset = 0, insetLeft = inset, insetRight = inset, insetBottom = inset, insetTop = inset } = style, rest = __rest(style, ["stroke", "radius", "radiusTopLeft", "radiusTopRight", "radiusBottomRight", "radiusBottomLeft", "innerRadius", "innerRadiusTopLeft", "innerRadiusTopRight", "innerRadiusBottomRight", "innerRadiusBottomLeft", "lineWidth", "inset", "insetLeft", "insetRight", "insetBottom", "insetTop"]);
        const { color = defaultColor, opacity } = value;
        // Extended style, which is not supported by native g shape,
        // should apply at first.
        const extendedStyle = {
            radius,
            radiusTopLeft: first ? radiusTopLeft : innerRadiusTopLeft,
            radiusTopRight: first ? radiusTopRight : innerRadiusTopRight,
            radiusBottomRight: last ? radiusBottomRight : innerRadiusBottomRight,
            radiusBottomLeft: last ? radiusBottomLeft : innerRadiusBottomLeft,
            inset,
            insetLeft,
            insetRight,
            insetBottom,
            insetTop,
        };
        return ((0, selection_1.select)(rect(points, value, coordinate, extendedStyle))
            .call(utils_1.applyStyle, defaults)
            .style(colorAttribute, color)
            .style((0, utils_1.toOpacityKey)(options), opacity)
            .style('lineWidth', lineWidth)
            .style('stroke', stroke === undefined ? color : stroke)
            // shape.style has higher priority.
            .call(utils_1.applyStyle, rest)
            .node());
    };
};
exports.Color = Color;
// @todo Should Shape have default animations using for ordinal scale?
exports.Color.props = {
    defaultEnterAnimation: 'scaleInY',
    defaultUpdateAnimation: 'morphing',
    defaultExitAnimation: 'fadeOut',
};
