// @antv/g2/lib/api/mark/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mark = exports.MarkNode = void 0;
const mark_1 = require("./mark");
var base_1 = require("./base");
Object.defineProperty(exports, "MarkNode", { enumerable: true, get: function () { return base_1.MarkNode; } });
exports.mark = {
    mark: mark_1.HOMMark,
    interval: mark_1.Interval,
    rect: mark_1.Rect,
    point: mark_1.Point,
    area: mark_1.Area,
    line: mark_1.Line,
    cell: mark_1.Cell,
    vector: mark_1.Vector,
    link: mark_1.Link,
    polygon: mark_1.Polygon,
    image: mark_1.Image,
    text: mark_1.Text,
    box: mark_1.Box,
    lineX: mark_1.LineX,
    lineY: mark_1.LineY,
    range: mark_1.Range,
    rangeX: mark_1.RangeX,
    rangeY: mark_1.RangeY,
    connector: mark_1.Connector,
    sankey: mark_1.Sankey,
    treemap: mark_1.Treemap,
    boxplot: mark_1.Boxplot,
    shape: mark_1.Shape,
    pack: mark_1.Pack,
    forceGraph: mark_1.ForceGraph,
    tree: mark_1.Tree,
    wordCloud: mark_1.WordCloud,
};
