// @antv/g2/lib/api/composition/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GeoPath = exports.GeoView = exports.TimingKeyframe = exports.RepeatMatrix = exports.FacetCircle = exports.FacetRect = exports.SpaceFlex = exports.SpaceLayer = exports.View = exports.CompositionNode = exports.composition = void 0;
const props_1 = require("../props");
const base_1 = require("./base");
Object.defineProperty(exports, "CompositionNode", { enumerable: true, get: function () { return base_1.CompositionNode; } });
const view_1 = require("./view");
Object.defineProperty(exports, "View", { enumerable: true, get: function () { return view_1.View; } });
const spaceLayer_1 = require("./spaceLayer");
Object.defineProperty(exports, "SpaceLayer", { enumerable: true, get: function () { return spaceLayer_1.SpaceLayer; } });
const spaceFlex_1 = require("./spaceFlex");
Object.defineProperty(exports, "SpaceFlex", { enumerable: true, get: function () { return spaceFlex_1.SpaceFlex; } });
const facetRect_1 = require("./facetRect");
Object.defineProperty(exports, "FacetRect", { enumerable: true, get: function () { return facetRect_1.FacetRect; } });
const facetCircle_1 = require("./facetCircle");
Object.defineProperty(exports, "FacetCircle", { enumerable: true, get: function () { return facetCircle_1.FacetCircle; } });
const repeatMatrix_1 = require("./repeatMatrix");
Object.defineProperty(exports, "RepeatMatrix", { enumerable: true, get: function () { return repeatMatrix_1.RepeatMatrix; } });
const timingKeyframe_1 = require("./timingKeyframe");
Object.defineProperty(exports, "TimingKeyframe", { enumerable: true, get: function () { return timingKeyframe_1.TimingKeyframe; } });
const geoView_1 = require("./geoView");
Object.defineProperty(exports, "GeoView", { enumerable: true, get: function () { return geoView_1.GeoView; } });
const geoPath_1 = require("./geoPath");
Object.defineProperty(exports, "GeoPath", { enumerable: true, get: function () { return geoPath_1.GeoPath; } });
exports.composition = {
    view: view_1.View,
    spaceLayer: spaceLayer_1.SpaceLayer,
    spaceFlex: spaceFlex_1.SpaceFlex,
    facetRect: facetRect_1.FacetRect,
    facetCircle: facetCircle_1.FacetCircle,
    repeatMatrix: repeatMatrix_1.RepeatMatrix,
    timingKeyframe: timingKeyframe_1.TimingKeyframe,
    geoView: geoView_1.GeoView,
    geoPath: geoPath_1.GeoPath,
};
/**
 * Define composition node api for composition node dynamically,
 * which can avoid circular dependency.
 * @todo Remove view as composition.
 */
for (const Ctor of Object.values(exports.composition)) {
    (0, props_1.defineProps)((0, props_1.nodeProps)(exports.composition))(Ctor);
}
