// @antv/g2/lib/composition/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimingKeyframe = exports.FacetCircle = exports.RepeatMatrix = exports.FacetRect = exports.Mark = exports.View = exports.SpaceFlex = exports.SpaceLayer = void 0;
var spaceLayer_1 = require("./spaceLayer");
Object.defineProperty(exports, "SpaceLayer", { enumerable: true, get: function () { return spaceLayer_1.SpaceLayer; } });
var spaceFlex_1 = require("./spaceFlex");
Object.defineProperty(exports, "SpaceFlex", { enumerable: true, get: function () { return spaceFlex_1.SpaceFlex; } });
var view_1 = require("./view");
Object.defineProperty(exports, "View", { enumerable: true, get: function () { return view_1.View; } });
var mark_1 = require("./mark");
Object.defineProperty(exports, "Mark", { enumerable: true, get: function () { return mark_1.Mark; } });
var facetRect_1 = require("./facetRect");
Object.defineProperty(exports, "FacetRect", { enumerable: true, get: function () { return facetRect_1.FacetRect; } });
var repeatMatrix_1 = require("./repeatMatrix");
Object.defineProperty(exports, "RepeatMatrix", { enumerable: true, get: function () { return repeatMatrix_1.RepeatMatrix; } });
var facetCircle_1 = require("./facetCircle");
Object.defineProperty(exports, "FacetCircle", { enumerable: true, get: function () { return facetCircle_1.FacetCircle; } });
var timingKeyframe_1 = require("./timingKeyframe");
Object.defineProperty(exports, "TimingKeyframe", { enumerable: true, get: function () { return timingKeyframe_1.TimingKeyframe; } });
