// @antv/gui/lib/ui/checkbox/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CHECKED_SHAPE_STYLE = exports.CHECKBOX_RECT_STYLE = exports.LABEL_TEXT_STYLE = void 0;
var tslib_1 = require("tslib");
var util_1 = require("../../util");
// 默认文本样式
exports.LABEL_TEXT_STYLE = tslib_1.__assign(tslib_1.__assign({}, util_1.TEXT_INHERITABLE_PROPS), { fill: 'rgba(0,0,0,0.45)', fontSize: 10, textAlign: 'start', textBaseline: 'middle', overflow: 'clip' });
exports.CHECKBOX_RECT_STYLE = {
    default: {
        width: 12,
        height: 12,
        radius: 2,
        stroke: '#dadada',
        lineWidth: 1,
        fill: '#ffffff',
        cursor: 'pointer',
    },
    selected: {
        width: 12,
        height: 12,
        radius: 2,
        stroke: '#3471F9',
        lineWidth: 1,
        fill: '#3471F9',
        cursor: 'pointer',
    },
};
var CHECKED_SHAPE_PATH = [
    ['M', 3, 6],
    ['L', '5', '8.5'],
    ['L', '8.5', '4'],
];
exports.CHECKED_SHAPE_STYLE = {
    path: CHECKED_SHAPE_PATH,
    lineWidth: 1,
    cursor: 'pointer',
};
