// @antv/gui/lib/ui/navigator/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Navigator = void 0;
var tslib_1 = require("tslib");
var g_1 = require("@antv/g");
var util_1 = require("@antv/util");
var gui_1 = require("../../core/gui");
var util_2 = require("../../util");
var symbol_1 = require("../marker/symbol");
var NAVIGATOR_DEFAULT_CFG = {
    effect: 'linear',
    duration: 200,
    orient: 'horizontal',
    initPage: 0,
    loop: false,
    buttonPath: (0, symbol_1.button)(0, 0, 6),
    buttonFill: 'black',
    buttonSize: 12,
    buttonCursor: 'pointer',
    formatter: function (curr, total) { return "".concat(curr, "/").concat(total); },
    pageNumFontSize: 12,
    pageNumFill: 'black',
    pageNumTextAlign: 'start',
    pageNumTextBaseline: 'middle',
    controllerPadding: 5,
    controllerSpacing: 5,
};
var CLASS_NAMES = (0, util_2.classNames)({
    prevBtnGroup: 'prev-btn-group',
    prevBtn: 'prev-btn',
    nextBtnGroup: 'next-btn-group',
    nextBtn: 'next-btn',
    pageInfoGroup: 'page-info-group',
    pageInfo: 'page-info',
    playWindow: 'play-window',
    contentGroup: 'content-group',
    controller: 'controller',
    clipPath: 'clip-path',
}, 'navigator');
var Navigator = /** @class */ (function (_super) {
    tslib_1.__extends(Navigator, _super);
    function Navigator(options) {
        var _this = _super.call(this, (0, util_2.deepAssign)({}, { style: NAVIGATOR_DEFAULT_CFG }, options)) || this;
        _this.finishedPromise = null;
        _this.resolveFinishedPromise = null;
        _this.playState = 'idle';
        _this.contentGroup = _this.appendChild(new g_1.Group({ class: CLASS_NAMES.contentGroup.name }));
        _this.playWindow = _this.contentGroup.appendChild(new g_1.Group({ class: CLASS_NAMES.playWindow.name }));
        _this.innerCurrPage = _this.defaultPage;
        return _this;
    }
    Object.defineProperty(Navigator.prototype, "defaultPage", {
        get: function () {
            var _a = this.attributes.initPage, initPage = _a === void 0 ? 0 : _a;
            return (0, util_1.clamp)(initPage, 0, Math.max(this.pageViews.length - 1, 0));
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Navigator.prototype, "pageViews", {
        get: function () {
            return this.playWindow.children;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Navigator.prototype, "controllerShape", {
        // todo fixme
        get: function () {
            return this.totalPages > 1 ? { width: 55, height: 0 } : { width: 0, height: 0 };
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Navigator.prototype, "pageShape", {
        get: function () {
            var pageViews = this.pageViews;
            var _a = tslib_1.__read((0, util_2.transpose)(pageViews.map(function (pageView) {
                var _a = pageView.getBBox(), width = _a.width, height = _a.height;
                return [width, height];
            })).map(function (arr) { return Math.max.apply(Math, tslib_1.__spreadArray([], tslib_1.__read(arr), false)); }), 2), maxWidth = _a[0], maxHeight = _a[1];
            var _b = this.attributes, _c = _b.pageWidth, pageWidth = _c === void 0 ? maxWidth : _c, _d = _b.pageHeight, pageHeight = _d === void 0 ? maxHeight : _d;
            return { pageWidth: pageWidth, pageHeight: pageHeight };
        },
        enumerable: false,
        configurable: true
    });
    Navigator.prototype.getContainer = function () {
        return this.playWindow;
    };
    Object.defineProperty(Navigator.prototype, "totalPages", {
        get: function () {
            return this.pageViews.length;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Navigator.prototype, "currPage", {
        get: function () {
            return this.innerCurrPage;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(Navigator.prototype, "finished", {
        get: function () {
            var _this = this;
            var _a;
            if (!this.finishedPromise) {
                this.finishedPromise = new Promise(function (resolve) {
                    _this.resolveFinishedPromise = function () {
                        _this.finishedPromise = null;
                        resolve(_this.currPage);
                    };
                });
            }
            if (this.playState === 'idle')
                (_a = this.resolveFinishedPromise) === null || _a === void 0 ? void 0 : _a.call(this);
            return this.finishedPromise;
        },
        enumerable: false,
        configurable: true
    });
    Navigator.prototype.getBBox = function () {
        var _a = _super.prototype.getBBox.call(this), x = _a.x, y = _a.y;
        var controllerShape = this.controllerShape;
        var _b = this.pageShape, pageWidth = _b.pageWidth, pageHeight = _b.pageHeight;
        return new util_2.BBox(x, y, pageWidth + controllerShape.width, pageHeight);
    };
    Navigator.prototype.goTo = function (pageNum) {
        var _this = this;
        var _a = this.attributes, duration = _a.duration, effect = _a.effect;
        var _b = this, currPage = _b.currPage, playState = _b.playState, finished = _b.finished, playWindow = _b.playWindow, pageViews = _b.pageViews;
        if (playState !== 'idle' || pageNum < 0 || pageViews.length <= 0 || pageNum >= pageViews.length)
            return { finished: this.finished };
        pageViews[currPage].setLocalPosition(0, 0);
        this.prepareFollowingPage(pageNum);
        var animateCfg = { duration: duration, easing: effect, fill: 'both' };
        var _c = tslib_1.__read(this.getFollowingPageDiff(pageNum), 2), dx = _c[0], dy = _c[1];
        this.playState = 'running';
        Promise.all([
            playWindow.animate([{ transform: "translate(0, 0)" }, { transform: "translate(".concat(-dx, ", ").concat(-dy, ")") }], animateCfg),
        ].map(function (ani) { return ani === null || ani === void 0 ? void 0 : ani.finished; })).then(function () {
            var _a;
            _this.innerCurrPage = pageNum;
            _this.playState = 'idle';
            _this.setVisiblePages([pageNum]);
            _this.updatePageInfo();
            (_a = _this.resolveFinishedPromise) === null || _a === void 0 ? void 0 : _a.call(_this);
        });
        return { finished: this.finished };
    };
    Navigator.prototype.prev = function () {
        var loop = this.attributes.loop;
        var pages = this.pageViews.length;
        var page = this.currPage;
        if (!loop && page <= 0)
            return { finished: this.finished };
        var following = loop ? (page - 1 + pages) % pages : (0, util_1.clamp)(page - 1, 0, pages);
        return this.goTo(following);
    };
    Navigator.prototype.next = function () {
        var loop = this.attributes.loop;
        var pages = this.pageViews.length;
        var page = this.currPage;
        if (!loop && page >= pages - 1)
            return { finished: this.finished };
        var following = loop ? (page + 1) % pages : (0, util_1.clamp)(page + 1, 0, pages);
        return this.goTo(following);
    };
    Navigator.prototype.renderClipPath = function (container) {
        var _a = this.pageShape, pageWidth = _a.pageWidth, pageHeight = _a.pageHeight;
        this.clipPath = container.maybeAppendByClassName(CLASS_NAMES.clipPath, 'rect').styles({
            width: pageWidth,
            height: pageHeight,
        });
        this.contentGroup.attr('clipPath', this.clipPath.node());
    };
    Navigator.prototype.setVisiblePages = function (pages) {
        this.playWindow.children.forEach(function (page, index) {
            if (pages.includes(index))
                page.attr('visibility', 'visible');
            else
                page.attr('visibility', 'hidden');
        });
    };
    Navigator.prototype.adjustControllerLayout = function () {
        var _a = this, prevBtn = _a.prevBtnGroup, nextBtn = _a.nextBtnGroup, pageNum = _a.pageInfoGroup;
        var _b = this.attributes, orient = _b.orient, padding = _b.controllerPadding;
        var _c = pageNum.getBBox(), pW = _c.width, pH = _c.height;
        var _d = tslib_1.__read(orient === 'horizontal' ? [-180, 0] : [-90, 90], 2), r1 = _d[0], r2 = _d[1];
        prevBtn.setLocalEulerAngles(r1);
        nextBtn.setLocalEulerAngles(r2);
        var _e = prevBtn.getBBox(), bpW = _e.width, bpH = _e.height;
        var _f = nextBtn.getBBox(), bnW = _f.width, bnH = _f.height;
        var maxWidth = Math.max(bpW, pW, bnW);
        var _g = orient === 'horizontal'
            ? {
                offset: [
                    [0, 0],
                    [bpW / 2 + padding, 0],
                    [bpW + pW + padding * 2, 0],
                ],
                textAlign: 'start',
            }
            : {
                offset: [
                    [maxWidth / 2, -bpH - padding],
                    [maxWidth / 2, 0],
                    [maxWidth / 2, bnH + padding],
                ],
                textAlign: 'center',
            }, _h = tslib_1.__read(_g.offset, 3), _j = tslib_1.__read(_h[0], 2), o1x = _j[0], o1y = _j[1], _k = tslib_1.__read(_h[1], 2), o2x = _k[0], o2y = _k[1], _l = tslib_1.__read(_h[2], 2), o3x = _l[0], o3y = _l[1], textAlign = _g.textAlign;
        var pageNumText = pageNum.querySelector('text');
        pageNumText && (pageNumText.style.textAlign = textAlign);
        prevBtn.setLocalPosition(o1x, o1y);
        pageNum.setLocalPosition(o2x, o2y);
        nextBtn.setLocalPosition(o3x, o3y);
    };
    Navigator.prototype.updatePageInfo = function () {
        var _a;
        var _b = this, currPage = _b.currPage, pageViews = _b.pageViews, formatter = _b.attributes.formatter;
        if (pageViews.length < 2)
            return;
        (_a = this.pageInfoGroup.querySelector(CLASS_NAMES.pageInfo.class)) === null || _a === void 0 ? void 0 : _a.attr('text', formatter(currPage + 1, pageViews.length));
        this.adjustControllerLayout();
    };
    Navigator.prototype.getFollowingPageDiff = function (pageNum) {
        var currPage = this.currPage;
        if (currPage === pageNum)
            return [0, 0];
        var orient = this.attributes.orient;
        var _a = this.pageShape, pageWidth = _a.pageWidth, pageHeight = _a.pageHeight;
        var sign = pageNum < currPage ? -1 : 1;
        return orient === 'horizontal' ? [sign * pageWidth, 0] : [0, sign * pageHeight];
    };
    Navigator.prototype.prepareFollowingPage = function (pageNum) {
        var _a = this, currPage = _a.currPage, pageViews = _a.pageViews;
        this.setVisiblePages([pageNum, currPage]);
        if (pageNum !== currPage) {
            var _b = tslib_1.__read(this.getFollowingPageDiff(pageNum), 2), dx = _b[0], dy = _b[1];
            pageViews[pageNum].setLocalPosition(dx, dy);
        }
    };
    Navigator.prototype.renderController = function (container) {
        var _this = this;
        var spacing = this.attributes.controllerSpacing;
        var _a = this.pageShape, pageWidth = _a.pageWidth, pageHeight = _a.pageHeight;
        var visible = this.pageViews.length >= 2;
        var group = container
            .maybeAppendByClassName(CLASS_NAMES.controller, 'g')
            .style('visibility', visible ? 'visible' : 'hidden');
        if (!visible)
            return;
        var _b = tslib_1.__read((0, util_2.subObjects)(this.attributes, ['button', 'pageNum']), 2), style = _b[0], textStyle = _b[1];
        var _c = tslib_1.__read((0, util_2.styleSeparator)(style), 2), _d = _c[0], groupStyle = _c[1], size = _d.size, pathStyle = tslib_1.__rest(_d, ["size"]);
        var prevBtnGroup = group.maybeAppendByClassName(CLASS_NAMES.prevBtnGroup, 'g').styles(groupStyle);
        this.prevBtnGroup = prevBtnGroup.node();
        var prevBtn = prevBtnGroup.maybeAppendByClassName(CLASS_NAMES.prevBtn, 'path');
        var nextBtnGroup = group.maybeAppendByClassName(CLASS_NAMES.nextBtnGroup, 'g').styles(groupStyle);
        this.nextBtnGroup = nextBtnGroup.node();
        var nextBtn = nextBtnGroup.maybeAppendByClassName(CLASS_NAMES.nextBtn, 'path');
        [prevBtn, nextBtn].forEach(function (btn) {
            btn.styles(tslib_1.__assign(tslib_1.__assign({}, pathStyle), { transformOrigin: 'center' }));
            (0, util_2.scaleToPixel)(btn.node(), size, true);
        });
        var pageInfoGroup = group.maybeAppendByClassName(CLASS_NAMES.pageInfoGroup, 'g');
        this.pageInfoGroup = pageInfoGroup.node();
        pageInfoGroup
            .maybeAppendByClassName(CLASS_NAMES.pageInfo, 'text')
            .styles(tslib_1.__assign(tslib_1.__assign({}, util_2.TEXT_INHERITABLE_PROPS), textStyle));
        this.updatePageInfo();
        // put it on the right side of the container
        group.node().setLocalPosition(pageWidth + spacing, pageHeight / 2);
        // add event
        prevBtnGroup.on('click', function () {
            _this.prev();
        });
        nextBtnGroup.on('click', function () {
            _this.next();
        });
    };
    Navigator.prototype.render = function (attributes, container) {
        /**
         * container
         *  |- contentGroup (with clip path)
         *    |- playWindow (with animation)
         *      |- pages
         *  |- clipPath
         */
        var containerSelection = (0, util_2.select)(container);
        this.renderClipPath(containerSelection);
        this.renderController(containerSelection);
        this.setVisiblePages([this.defaultPage]);
        this.goTo(this.defaultPage);
    };
    Navigator.prototype.bindEvents = function () {
        var _this = this;
        var render = (0, util_1.debounce)(function () { return _this.render(_this.attributes, _this); }, 50);
        this.playWindow.addEventListener(g_1.ElementEvent.INSERTED, render);
        this.playWindow.addEventListener(g_1.ElementEvent.REMOVED, render);
    };
    return Navigator;
}(gui_1.GUI));
exports.Navigator = Navigator;
