// @antv/gui/esm/ui/slider/handle.js
import { __assign, __extends, __read, __rest } from "tslib";
import { GUI } from '../../core/gui';
import { classNames, select, subObject, deepAssign, ifShow, styleSeparator } from '../../util';
import { createComponent } from '../../util/create';
import { HANDLE_DEFAULT_CFG, HANDLE_ICON_DEFAULT_CFG, HANDLE_LABEL_DEFAULT_CFG } from './constant';
var CLASS_NAMES = classNames({
    labelGroup: 'label-group',
    label: 'label',
    icon: 'icon',
    iconRect: 'icon-rect',
    iconLine: 'icon-line',
}, 'handle');
var HandleIcon = createComponent({
    render: function (attributes, container) {
        var _a = attributes.size, size = _a === void 0 ? 10 : _a, _b = attributes.radius, radius = _b === void 0 ? size / 4 : _b, orient = attributes.orient, className = attributes.class, iconStyle = __rest(attributes, ["size", "radius", "orient", "class"]);
        // 默认手柄
        var width = size;
        var height = width * 2.4;
        var rect = select(container)
            .maybeAppendByClassName(CLASS_NAMES.iconRect, 'rect')
            .styles(__assign(__assign({}, iconStyle), { width: width, height: height, radius: radius, x: -width / 2, y: -height / 2 }));
        var x1 = (1 / 3) * width;
        var x2 = (2 / 3) * width;
        var y1 = (1 / 4) * height;
        var y2 = (3 / 4) * height;
        rect.maybeAppendByClassName("".concat(CLASS_NAMES.iconLine, "-1"), 'line').styles(__assign({ x1: x1, x2: x1, y1: y1, y2: y2 }, iconStyle));
        rect.maybeAppendByClassName("".concat(CLASS_NAMES.iconLine, "-2"), 'line').styles(__assign({ x1: x2, x2: x2, y1: y1, y2: y2 }, iconStyle));
        rect.node().setOrigin(width / 2, height / 2);
        if (orient === 'vertical')
            container.setLocalEulerAngles(90);
        else
            container.setLocalEulerAngles(0);
    },
});
var Handle = /** @class */ (function (_super) {
    __extends(Handle, _super);
    function Handle(config) {
        return _super.call(this, deepAssign({}, { style: HANDLE_DEFAULT_CFG }, config)) || this;
    }
    Handle.prototype.renderLabel = function (container) {
        var _this = this;
        var showLabel = this.attributes.showLabel;
        var style = subObject(this.attributes, 'label');
        var _a = __read(styleSeparator(style, []), 2), labelStyle = _a[0], groupStyle = _a[1];
        var labelGroup = select(container).maybeAppendByClassName(CLASS_NAMES.labelGroup, 'g').styles(groupStyle);
        ifShow(!!showLabel, labelGroup, function (group) {
            _this.label = group
                .maybeAppendByClassName(CLASS_NAMES.label, 'text')
                .styles(__assign(__assign({}, HANDLE_LABEL_DEFAULT_CFG), labelStyle));
            /** avoid trigger event on label */
            _this.label.on('mousedown', function (e) {
                e.stopPropagation();
            });
            _this.label.on('touchstart', function (e) {
                e.stopPropagation();
            });
        });
    };
    Handle.prototype.renderIcon = function (container) {
        var _a = this.attributes, _b = _a.iconShape, iconShape = _b === void 0 ? function () {
            return new HandleIcon({ style: iconStyle });
        } : _b, orient = _a.orient;
        var iconStyle = __assign(__assign({ orient: orient }, HANDLE_ICON_DEFAULT_CFG), subObject(this.attributes, 'icon'));
        this.icon = select(container).maybeAppendByClassName(CLASS_NAMES.icon, iconShape).update(iconStyle);
    };
    Handle.prototype.render = function (attributes, container) {
        this.renderIcon(container);
        this.renderLabel(container);
    };
    return Handle;
}(GUI));
export { Handle };
