// @antv/g2/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.register = exports.CompositionNode = exports.MarkNode = exports.Chart = exports.createLibrary = exports.AREA_CLASS_NAME = exports.LABEL_CLASS_NAME = exports.COMPONENT_CLASS_NAME = exports.PLOT_CLASS_NAME = exports.VIEW_CLASS_NAME = exports.ELEMENT_CLASS_NAME = exports.LABEL_LAYER_CLASS_NAME = exports.MAIN_LAYER_CLASS_NAME = exports.renderToMountedElement = exports.render = void 0;
var runtime_1 = require("./runtime");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return runtime_1.render; } });
Object.defineProperty(exports, "renderToMountedElement", { enumerable: true, get: function () { return runtime_1.renderToMountedElement; } });
Object.defineProperty(exports, "MAIN_LAYER_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.MAIN_LAYER_CLASS_NAME; } });
Object.defineProperty(exports, "LABEL_LAYER_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.LABEL_LAYER_CLASS_NAME; } });
Object.defineProperty(exports, "ELEMENT_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.ELEMENT_CLASS_NAME; } });
Object.defineProperty(exports, "VIEW_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.VIEW_CLASS_NAME; } });
Object.defineProperty(exports, "PLOT_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.PLOT_CLASS_NAME; } });
Object.defineProperty(exports, "COMPONENT_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.COMPONENT_CLASS_NAME; } });
Object.defineProperty(exports, "LABEL_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.LABEL_CLASS_NAME; } });
Object.defineProperty(exports, "AREA_CLASS_NAME", { enumerable: true, get: function () { return runtime_1.AREA_CLASS_NAME; } });
var stdlib_1 = require("./stdlib");
Object.defineProperty(exports, "createLibrary", { enumerable: true, get: function () { return stdlib_1.createLibrary; } });
var api_1 = require("./api");
Object.defineProperty(exports, "Chart", { enumerable: true, get: function () { return api_1.Chart; } });
Object.defineProperty(exports, "MarkNode", { enumerable: true, get: function () { return api_1.MarkNode; } });
Object.defineProperty(exports, "CompositionNode", { enumerable: true, get: function () { return api_1.CompositionNode; } });
Object.defineProperty(exports, "register", { enumerable: true, get: function () { return api_1.register; } });
__exportStar(require("./spec"), exports);
