// @antv/gui/lib/ui/axis/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CLASS_NAMES = exports.HELIX_DEFAULT_OPTIONS = exports.ARC_DEFAULT_OPTIONS = exports.AXIS_BASE_DEFAULT_CFG = void 0;
var g_1 = require("@antv/g");
var util_1 = require("@antv/util");
var util_2 = require("../../util");
exports.AXIS_BASE_DEFAULT_CFG = {
    style: {
        data: [],
        animate: {
            duration: 100,
            easing: 'ease-in-out-sine',
            fill: 'both',
        },
        dataThreshold: 100,
        showTitle: true,
        titlePosition: 'lb',
        titleFill: 'black',
        titleFontSize: 12,
        titleTextAlign: 'center',
        titleSpacing: 0,
        titleTextBaseline: 'middle',
        showLine: true,
        lineLineWidth: 1,
        lineStroke: 'black',
        showArrow: true,
        lineArrow: new g_1.Path({
            style: {
                path: 'M 10,10 L -10,0 L 10,-10 L0,0 L10,10Z',
                anchor: '0.5 0.5',
                fill: 'black',
                transformOrigin: 'center',
            },
        }),
        lineArrowOffset: 15,
        lineArrowSize: 10,
        showTick: true,
        tickDirection: 'positive',
        tickStroke: 'black',
        tickLength: 5,
        tickLineWidth: 1,
        tickStrokeOpacity: 0.65,
        showLabel: true,
        labelSpacing: 0,
        labelDirection: 'positive',
        labelAlign: 'parallel',
        labelTransforms: [
        // { type: 'rotate', optionalAngles: [0, 45, 90] },
        // { type: 'ellipsis', suffix: '...', minLength: 14, maxLength: 160 },
        // { type: 'hide' },
        ],
        showGrid: true,
        gridType: 'segment',
        gridConnect: 'line',
        gridDirection: 'positive',
        gridLength: 0,
        gridControlAngles: [],
    },
};
exports.ARC_DEFAULT_OPTIONS = (0, util_1.deepMix)({}, exports.AXIS_BASE_DEFAULT_CFG, {
    style: {
        type: 'arc',
    },
});
exports.HELIX_DEFAULT_OPTIONS = (0, util_1.deepMix)({}, exports.AXIS_BASE_DEFAULT_CFG, {
    style: {},
});
exports.CLASS_NAMES = (0, util_2.classNames)({
    mainGroup: 'main-group',
    gridGroup: 'grid-group',
    grid: 'grid',
    lineGroup: 'line-group',
    line: 'line',
    tickGroup: 'tick-group',
    tick: 'tick',
    tickItem: 'tick-item',
    labelGroup: 'label-group',
    label: 'label',
    labelItem: 'label-item',
    titleGroup: 'title-group',
    title: 'title',
    lineFirst: 'line-first',
    lineSecond: 'line-second',
}, 'axis');
