// @antv/gui/lib/ui/legend/continuous/ribbon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ribbon = void 0;
var tslib_1 = require("tslib");
var g_1 = require("@antv/g");
var util_1 = require("@antv/util");
var util_2 = require("../../../util");
var utils_1 = require("../utils");
var utils_2 = require("./utils");
var CLASS_NAMES = (0, util_2.classNames)({
    trackGroup: 'background-group',
    track: 'background',
    selectionGroup: 'ribbon-group',
    selection: 'ribbon',
    clipPath: 'clip-path',
}, 'ribbon');
var DEFAULT_RIBBON_CFG = {
    type: 'color',
    orient: 'horizontal',
    size: 30,
    range: [0, 1],
    len: 200,
    block: false,
    partition: [],
    color: ['#fff', '#000'],
    trackFill: '#e5e5e5',
};
function getShape(cfg) {
    var orient = cfg.orient, size = cfg.size, len = cfg.len;
    return (0, utils_1.ifHorizontal)(orient, [len, size], [size, len]);
}
function getTrackPath(cfg) {
    var type = cfg.type;
    var _a = tslib_1.__read(getShape(cfg), 2), cw = _a[0], ch = _a[1];
    if (type === 'size') {
        return [['M', 0, ch], ['L', 0 + cw, 0], ['L', 0 + cw, ch], ['Z']];
    }
    return [['M', 0, ch], ['L', 0, 0], ['L', 0 + cw, 0], ['L', 0 + cw, ch], ['Z']];
}
function getSelectionPath(cfg) {
    return getTrackPath(cfg);
}
function getColor(cfg) {
    var _a = cfg, orient = _a.orient, color = _a.color, block = _a.block, partition = _a.partition;
    var colors;
    if ((0, util_1.isFunction)(color)) {
        var len = 20;
        colors = new Array(len).fill(0).map(function (_, index, arr) { return color(index / (arr.length - 1)); });
    }
    else
        colors = color;
    var count = colors.length;
    var genericColor = colors.map(function (c) { return (0, g_1.parseColor)(c).toString(); });
    if (!count)
        return '';
    if (count === 1)
        return genericColor[0];
    if (block)
        return (0, utils_2.getBlockColor)(partition, genericColor, orient);
    return genericColor.reduce(function (r, c, idx) { return (r += " ".concat(idx / (count - 1), ":").concat(c)); }, "l(".concat((0, utils_1.ifHorizontal)(orient, '0', '270'), ")"));
}
function getClipPath(cfg) {
    var orient = cfg.orient, range = cfg.range;
    if (!range)
        return [];
    var _a = tslib_1.__read(getShape(cfg), 2), width = _a[0], height = _a[1];
    var _b = tslib_1.__read(range, 2), st = _b[0], et = _b[1];
    var x = (0, utils_1.ifHorizontal)(orient, st * width, 0);
    var y = (0, utils_1.ifHorizontal)(orient, 0, st * height);
    var w = (0, utils_1.ifHorizontal)(orient, et * width, width);
    var h = (0, utils_1.ifHorizontal)(orient, height, et * height);
    return [['M', x, y], ['L', x, h], ['L', w, h], ['L', w, y], ['Z']];
}
function renderTrack(container, cfg, style) {
    container.maybeAppendByClassName(CLASS_NAMES.track, 'path').styles(tslib_1.__assign({ path: getTrackPath(cfg) }, style));
}
function renderSelection(container, cfg, style) {
    var fill = getColor(cfg);
    var ribbon = container
        .maybeAppendByClassName(CLASS_NAMES.selection, 'path')
        .styles(tslib_1.__assign({ path: getSelectionPath(cfg), fill: fill }, style));
    var clipPath = ribbon
        .maybeAppendByClassName(CLASS_NAMES.clipPath, 'path')
        .styles({ path: getClipPath(cfg) })
        .node();
    ribbon.style('clip-path', clipPath);
}
exports.Ribbon = (0, util_2.createComponent)({
    render: function (attribute, container) {
        var _a = tslib_1.__read((0, util_2.subObjects)(attribute, ['selection', 'track']), 2), selectionStyle = _a[0], trackStyle = _a[1];
        var trackGroup = (0, util_2.select)(container).maybeAppendByClassName(CLASS_NAMES.trackGroup, 'g');
        renderTrack(trackGroup, attribute, trackStyle);
        /**
         * - ribbon group
         *  |- ribbon
         * - clip path
         */
        var ribbonGroup = (0, util_2.select)(container).maybeAppendByClassName(CLASS_NAMES.selectionGroup, 'g');
        renderSelection(ribbonGroup, attribute, selectionStyle);
    },
}, tslib_1.__assign({}, DEFAULT_RIBBON_CFG));
