// @antv/gui/lib/ui/sparkline/lines.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Lines = void 0;
var tslib_1 = require("tslib");
var g_1 = require("@antv/g");
var util_1 = require("@antv/util");
var util_2 = require("../../util");
var Lines = /** @class */ (function (_super) {
    tslib_1.__extends(Lines, _super);
    function Lines(_a) {
        var _this = this;
        var style = _a.style, rest = tslib_1.__rest(_a, ["style"]);
        _this = _super.call(this, (0, util_1.deepMix)({}, { type: 'lines', style: { width: 0, height: 0 } }, tslib_1.__assign({ style: style }, rest))) || this;
        _this.linesGroup = _this.appendChild(new g_1.Group({ name: 'lines' }));
        _this.areasGroup = _this.appendChild(new g_1.Group({ name: 'areas' }));
        _this.render();
        return _this;
    }
    Lines.prototype.render = function () {
        var _a = this.attributes, lines = _a.lines, areas = _a.areas;
        if (lines)
            this.createLines(lines);
        if (areas)
            this.createAreas(areas);
    };
    Lines.prototype.clear = function () {
        this.linesGroup.removeChildren();
        this.areasGroup.removeChildren();
    };
    Lines.prototype.update = function (cfg) {
        var lines = cfg.lines, areas = cfg.areas;
        this.clear();
        lines && this.setAttribute('lines', lines);
        areas && this.setAttribute('areas', areas);
        this.render();
    };
    Lines.prototype.createLines = function (lines) {
        (0, util_2.select)(this.linesGroup)
            .selectAll('line')
            .data(lines)
            .join(function (enter) {
            return enter.append('path').each(function (cfg) {
                (0, util_2.select)(this).styles(cfg);
            });
        }, function (update) {
            return update.each(function (cfg) {
                (0, util_2.select)(this).styles(cfg);
            });
        }, function (remove) { return remove.remove(); });
    };
    Lines.prototype.createAreas = function (areas) {
        (0, util_2.select)(this.linesGroup)
            .selectAll('area')
            .data(areas)
            .join(function (enter) {
            return enter.append('path').each(function (cfg) {
                (0, util_2.select)(this).styles(cfg);
            });
        }, function (update) {
            return update.each(function (cfg) {
                (0, util_2.select)(this).styles(cfg);
            });
        }, function (remove) { return remove.remove(); });
    };
    return Lines;
}(g_1.DisplayObject));
exports.Lines = Lines;
