// @antv/g2/lib/component/slider.js
"use strict";
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Slider = void 0;
const gui_1 = require("@antv/gui");
const d3_format_1 = require("d3-format");
const scale_1 = require("../utils/scale");
/**
 * Slider component.
 */
const Slider = (options) => {
    // do not pass size.
    const { orient, labelFormatter, size, style } = options, rest = __rest(options, ["orient", "labelFormatter", "size", "style"]);
    return ([scale], value, coordinate, theme) => {
        var _a;
        const { bbox } = value;
        const { x, y, width, height } = bbox;
        const { slider: sliderTheme = {} } = theme;
        const defaultFormatter = ((_a = scale.getFormatter) === null || _a === void 0 ? void 0 : _a.call(scale)) || ((v) => v);
        const formatter = typeof labelFormatter === 'string'
            ? (0, d3_format_1.format)(labelFormatter)
            : labelFormatter;
        return new gui_1.Slider({
            className: 'slider',
            style: Object.assign({}, sliderTheme, Object.assign(Object.assign({ x,
                y, trackLength: orient === 'horizontal' ? width : height, orient, formatter: (v) => {
                    const f = formatter || defaultFormatter;
                    // @todo Pass index to distinguish the left and the right value.
                    const tick = (0, scale_1.invert)(scale, v, true);
                    return f(tick);
                } }, style), rest)),
        });
    };
};
exports.Slider = Slider;
exports.Slider.props = {
    defaultPosition: 'bottom',
    defaultSize: 24,
    defaultOrder: 1,
};
